### PR TITLE
Add boolean literals and relational operators

### PR DESCRIPTION
**Pull request by akalicki**
Addresses #31.
 - add boolean literals
 - add unary `!` not operator
 - add binary relational operators
 - add full set of tests for all above 

cc @lillyfwang @alexandramedway @dannyech for review

### CODE REVIEW

File: compiler/ast.mli [hunk 1]
```diff
@@ -15,18 +15,25 @@ type op =    (* Operators *)
...
   | Int_lit of int               (* 42 *)
   | Float_lit of float           (* 42.0 *)
   | String_lit of string         (* "Hello, world" *)
+  | Bool_lit of bool             (* true *)
   | Id of string                 (* x *)
-  | Binop of expr * op * expr    (* a + b *)
-  | Unop of op * expr            (* -5 *)
   | Call of string * expr list   (* add(1, 2) *)
+  | Unop of op * expr            (* -5 *)
+  | Binop of expr * op * expr    (* a + b *)
```
Comment (lillyfwang) [inline compiler/ast.mli]: Why did you change the order here and in generator.ml? Is it a precedence thing?

Comment (akalicki) [inline compiler/ast.mli]: Oh whoops I was changing it to be consistent across all files / with the microc stuff, but it's still different in every file so fail.

File: test/compiler/_literal.ods [hunk 2]
```diff
@@ -1,6 +1,7 @@
 state 42
 state 42.1
 state "forty-two"
+
```
Comment (lillyfwang) [inline test/compiler/_literal.ods]: There is a [special circle in hell](http://www.buzzfeed.com/samir/things-that-will-drive-your-ocd-self-insane) for people like you....
#perfectionist

File: compiler/ast.mli [hunk 3]
```diff
@@ -15,18 +15,25 @@ type op =    (* Operators *)
   | Div      (* / *)
   | Mod      (* % *)
   | Pow      (* ** *)
+  | Not      (* ! *)
+  | Equal    (* == *)
```
Comment (alexandramedway) [inline compiler/ast.mli]: Equal -> Eq 
for consistency

File: compiler/ast.mli [hunk 4]
```diff
@@ -15,18 +15,25 @@ type op =    (* Operators *)
...
+  | Greater  (* > *)
+  | Geq      (* >= *)
 
 type expr =                      (* Expressions *)
   | Int_lit of int               (* 42 *)
   | Float_lit of float           (* 42.0 *)
   | String_lit of string         (* "Hello, world" *)
+  | Bool_lit of bool             (* true *)
   | Id of string                 (* x *)
-  | Binop of expr * op * expr    (* a + b *)
```
Comment (alexandramedway) [inline compiler/ast.mli]: Why did you move this?

File: compiler/generator.ml [hunk 5]
```diff
@@ -18,15 +18,23 @@ let txt_of_op = function
...
+  | Less -> "<"
+  | Leq -> "<="
+  | Greater -> ">"
+  | Geq -> ">="
 
 let rec txt_of_expr = function
   | Int_lit(i) -> string_of_int(i)
   | Float_lit(f) -> string_of_float(f)
   | String_lit(s) -> sprintf "\"%s\"" s
+  | Bool_lit(b) -> String.capitalize (string_of_bool(b))
```
Comment (alexandramedway) [inline compiler/generator.ml]: true -> True?

File: compiler/parser.mly [hunk 6]
```diff
@@ -48,9 +48,12 @@
 %token VOID_LITERAL
 
 /* Precedence and associativity of each operator */
+%left EQ NEQ
+%left LCAR LEQ RCAR GEQ
```
Comment (alexandramedway) [inline compiler/parser.mly]: LEQ and GEQ same precedence as EQ and NEQ no?

File: compiler/parser.mly [hunk 7]
```diff
@@ -77,18 +80,26 @@ stmt:
 
 expr:
   | literal                     { $1 }
-  | ID LPAREN args_opt RPAREN   { Call($1, $3)}
```
Comment (alexandramedway) [inline compiler/parser.mly]: Are you sure this should be moved? It was the second thing in most other asts, and I believe order matters

File: test/compiler/_relational.out [hunk 8]
```diff
@@ -0,0 +1,10 @@
+False
+True
+True
+True
+True
+True
+True
+True
+False
+False
```
Comment (alexandramedway) [inline test/compiler/_relational.out]: what about the print statements?

File: compiler/generator.ml (hunk 5 above)
Comment (akalicki) [inline compiler/generator.ml]: yep, in Python literals are `True` and `False`

File: compiler/parser.mly (hunk 6 above)
Comment (akalicki) [inline compiler/parser.mly]: I think EQ and NEQ have to have lower precedence, consider
```
true == 3 < 2
````
or
```
a < b == c < d
```

I pulled the precedence out of the C operator precedence table so it should be the same!

File: compiler/ast.mli (hunk 4 above)
Comment (akalicki) [inline compiler/ast.mli]: consistency with the other files. the order doesn't really matter unless it could get confused with things above it.

File: compiler/parser.mly (hunk 7 above)
Comment (akalicki) [inline compiler/parser.mly]: order only matters if there's things that could get confused with it. but our tests are still passing after the move so it shouldn't make a difference with what we have so far.

File: test/compiler/_relational.out (hunk 8 above)
Comment (akalicki) [inline test/compiler/_relational.out]: see `_relational.py` - this is just the expected output of running the compiled script.
the input file `_relational.ods` is compiled through our workflow, the python is checked against `_relational.py`, and then the output is checked against `_relational.out`.
